File: 3491-find-the-maximum-length-of-valid-subsequence-ii/3491-find-the-maximum-length-of-valid-subsequence-ii.py
class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        n = len(nums);
        dp = [[1] * k for _ in range(n)]

        max_length = 1

        for i in range(1, n):
            for j in range(i):
                mod = (nums[i] + nums[j]) % k
                dp[i][mod] = max(dp[i][mod], dp[j][mod] + 1)
                max_length = max(max_length, dp[i][mod])

        return max_length
        

File: 3491-find-the-maximum-length-of-valid-subsequence-ii/3491-find-the-maximum-length-of-valid-subsequence-ii.java
class Solution {
    public int maximumLength(int[] nums, int k) {
       int n = nums.length;
       int[][] dp = new int[n][k];

       int maxLength = 1;

       for(int i = 0;i < n; i++){
        Arrays.fill(dp[i], 1);

       }

       for(int i = 1; i < n; i++){
        for(int j = 0; j < i; j++){
            int mod = (nums[i] + nums[j]) % k;

            dp[i][mod] = Math.max(dp[i][mod], dp[j][mod] + 1);
            maxLength = Math.max(maxLength, dp[i][mod]);

        }
       } 

       return maxLength;
    }
}